### PR TITLE
common/perf_counters: don't let dec() go negative 

### PR DESCRIPTION
**Pull request by badone**
Fixes: https://tracker.ceph.com/issues/47932

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by badone** [comment]
jenkins test make check

**Comment by badone** [comment]
jenkins test make check

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
jenkins test make check

**Comment by batrick** [comment]
jenkins test api

**Comment by badone** [comment]
This is actually causing the test failure in unittest_rgw_dmclock_scheduler so I'll need to take a look. Probably missing a 'u' but need to confirm that. Setting DNM until I get time to look into it further.

**Comment by badone** [comment]
unittest_rgw_dmclock_scheduler seems to rely on undefined behaviour as follows.

Using breakpoints as follows (note: line numbers are slightly different
depending on whether this patch is present or not).

```
break src/ceph3/src/common/perf_counters.cc:178
  commands
    p idx
    p data.u64._M_i
    cont
  end
break src/ceph3/src/common/perf_counters.cc:194
  commands
    p idx
    p data.u64._M_i
    cont
  end
break src/test/rgw/test_rgw_dmclock_scheduler.cc:69
  commands
    cont
  end
```

With my patch the test fails with the following message.

```
/home/brad/working/src/ceph3/src/test/rgw/test_rgw_dmclock_scheduler.cc:69: Failure Expected equality of these values:
  0u
    Which is: 0
  counters(client_id::admin)->get(queue_counters::l_qlen)
    Which is: 1
```

Running it in gdb.

```
Starting program: /home/brad/working/src/ceph3/build/bin/unittest_rgw_dmclock_scheduler --gtest_filter=Queue.SyncRequest
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
[New Thread 0x7fffebe1f700 (LWP 3462252)]
[New Thread 0x7fffeb61e700 (LWP 3462253)]
[New Thread 0x7fffeae1d700 (LWP 3462254)]
[New Thread 0x7fffea61c700 (LWP 3462255)]

Thread 1 "unittest_rgw_dm" hit Breakpoint 7, ceph::common::PerfCounters::inc (this=0x5555558f2730, idx=427151, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:178
178	}
$158 = 427151
$159 = 1

Thread 1 "unittest_rgw_dm" hit Breakpoint 7, ceph::common::PerfCounters::inc (this=0x5555558f2730, idx=427151, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:178
178	}
$160 = 427151
$161 = 1

Thread 1 "unittest_rgw_dm" hit Breakpoint 8, ceph::common::PerfCounters::dec (this=0x5555558f2730, idx=427151, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:197
197	}
$162 = 427151
$163 = 0

Thread 1 "unittest_rgw_dm" hit Breakpoint 8, ceph::common::PerfCounters::dec (this=0x5555558f2730, idx=427151, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:197
197	}
$164 = 427151
$165 = 0

Thread 1 "unittest_rgw_dm" hit Breakpoint 7, ceph::common::PerfCounters::inc (this=0x5555558f2730, idx=427151, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:178
178	}
$166 = 427151
$167 = 1

Thread 1 "unittest_rgw_dm" hit Breakpoint 7, ceph::common::PerfCounters::inc (this=0x5555558f2730, idx=427151, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:178
178	}
$168 = 427151
$169 = 1

Thread 1 "unittest_rgw_dm" hit Breakpoint 7, ceph::common::PerfCounters::inc (this=0x5555558deae0, idx=427151, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:178
178	}
$170 = 427151
$171 = 1

Thread 1 "unittest_rgw_dm" hit Breakpoint 7, ceph::common::PerfCounters::inc (this=0x5555558deae0, idx=427151, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:178
178	}
$172 = 427151
$173 = 1

Thread 1 "unittest_rgw_dm" hit Breakpoint 8, ceph::common::PerfCounters::dec (this=0x5555558deae0, idx=427151, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:197
197	}
$174 = 427151
$175 = 0

Thread 1 "unittest_rgw_dm" hit Breakpoint 8, ceph::common::PerfCounters::dec (this=0x5555558deae0, idx=427151, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:197
197	}
$176 = 427151
$177 = 0

Thread 1 "unittest_rgw_dm" hit Breakpoint 7, ceph::common::PerfCounters::inc (this=0x5555558deae0, idx=427151, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:178
178	}
$178 = 427151
$179 = 1

Thread 1 "unittest_rgw_dm" hit Breakpoint 7, ceph::common::PerfCounters::inc (this=0x5555558deae0, idx=427151, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:178
178	}
$180 = 427151
$181 = 1

Thread 1 "unittest_rgw_dm" hit Breakpoint 9, rgw::dmclock::Queue_SyncRequest_Test::TestBody (this=0x55555593af10) at /home/brad/working/src/ceph3/src/test/rgw/test_rgw_dmclock_scheduler.cc:69
69	  EXPECT_EQ(0u, counters(client_id::admin)->get(queue_counters::l_qlen));
[Thread 0x7fffea61c700 (LWP 3462255) exited]
[Thread 0x7fffeae1d700 (LWP 3462254) exited]
[Thread 0x7fffeb61e700 (LWP 3462253) exited]
[Thread 0x7fffebe1f700 (LWP 3462252) exited]
[Inferior 1 (process 3462238) exited with code 01]
```

Running the original code (which passes) in gdb shows the following.

```
Starting program: /home/brad/working/src/ceph3/build/bin/unittest_rgw_dmclock_scheduler --gtest_filter=Queue.SyncRequest
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
[New Thread 0x7fffebe1f700 (LWP 3467061)]
[New Thread 0x7fffeb61e700 (LWP 3467062)]
[New Thread 0x7fffeae1d700 (LWP 3467063)]
[New Thread 0x7fffea61c700 (LWP 3467064)]

Thread 1 "unittest_rgw_dm" hit Breakpoint 5, ceph::common::PerfCounters::inc (this=0x5555558f2730, idx=427153, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:178
178	}
$17 = 427153
$18 = 1

Thread 1 "unittest_rgw_dm" hit Breakpoint 5, ceph::common::PerfCounters::inc (this=0x5555558f2730, idx=427154, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:178
178	}
$19 = 427154
$20 = 1

Thread 1 "unittest_rgw_dm" hit Breakpoint 6, ceph::common::PerfCounters::dec (this=0x5555558f2730, idx=427151, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:194
194	}
$21 = 427151
$22 = 18446744073709551615

Thread 1 "unittest_rgw_dm" hit Breakpoint 6, ceph::common::PerfCounters::dec (this=0x5555558f2730, idx=427152, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:194
194	}
$23 = 427152
$24 = 18446744073709551615

Thread 1 "unittest_rgw_dm" hit Breakpoint 5, ceph::common::PerfCounters::inc (this=0x5555558f2730, idx=427151, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:178
178	}
$25 = 427151
$26 = 0

Thread 1 "unittest_rgw_dm" hit Breakpoint 5, ceph::common::PerfCounters::inc (this=0x5555558f2730, idx=427152, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:178
178	}
$27 = 427152
$28 = 0

Thread 1 "unittest_rgw_dm" hit Breakpoint 5, ceph::common::PerfCounters::inc (this=0x5555558deae0, idx=427155, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:178
178	}
$29 = 427155
$30 = 1

Thread 1 "unittest_rgw_dm" hit Breakpoint 5, ceph::common::PerfCounters::inc (this=0x5555558deae0, idx=427156, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:178
178	}
$31 = 427156
$32 = 1

Thread 1 "unittest_rgw_dm" hit Breakpoint 6, ceph::common::PerfCounters::dec (this=0x5555558deae0, idx=427151, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:194
194	}
$33 = 427151
$34 = 18446744073709551615

Thread 1 "unittest_rgw_dm" hit Breakpoint 6, ceph::common::PerfCounters::dec (this=0x5555558deae0, idx=427152, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:194
194	}
$35 = 427152
$36 = 18446744073709551615

Thread 1 "unittest_rgw_dm" hit Breakpoint 5, ceph::common::PerfCounters::inc (this=0x5555558deae0, idx=427151, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:178
178	}
$37 = 427151
$38 = 0

Thread 1 "unittest_rgw_dm" hit Breakpoint 5, ceph::common::PerfCounters::inc (this=0x5555558deae0, idx=427152, amt=1) at /home/brad/working/src/ceph3/src/common/perf_counters.cc:178
178	}
$39 = 427152
$40 = 0

Thread 1 "unittest_rgw_dm" hit Breakpoint 7, rgw::dmclock::Queue_SyncRequest_Test::TestBody (this=0x55555593af10) at /home/brad/working/src/ceph3/src/test/rgw/test_rgw_dmclock_scheduler.cc:69
69	  EXPECT_EQ(1u, counters(client_id::admin)->get(queue_counters::l_res));
[Thread 0x7fffea61c700 (LWP 3467064) exited]
[Thread 0x7fffeae1d700 (LWP 3467063) exited]
[Thread 0x7fffeb61e700 (LWP 3467062) exited]
[Thread 0x7fffebe1f700 (LWP 3467061) exited]
[Inferior 1 (process 3467030) exited normally]
```

So the current test seems to rely on the counter going negative (which is UB
(implementation dependant).

@cbodley and @theanalyst mond taking a look and giving this PR a positive review if you agree the problem lies elsewhere?

File: src/common/perf_counters.cc [hunk 1]
```diff
@@ -190,7 +190,14 @@ void PerfCounters::dec(int idx, uint64_t amt)
...
   if (!(data.type & PERFCOUNTER_U64))
     return;
-  data.u64 -= amt;
+  if (amt > data.u64) { // would wrap
+    assert(false); // Assert in debug builds
+    lderr(m_cct) << "WARNING: PerfCounters::dec() " << data.u64 << " - "
+            << amt << " would wrap. Setting to 0" << dendl;
+    data.u64 = 0;
+  }
+  else
```
Comment (jdurgin) [inline src/common/perf_counters.cc:199]: nit: should have braces around the else clause since there are braces around the if clause

**Review by jdurgin — COMMENTED** [review]
lgtm once the unit tests are fixed

File: src/common/perf_counters.cc [hunk 2]
```diff
@@ -190,7 +190,14 @@ void PerfCounters::dec(int idx, uint64_t amt)
   ceph_assert(!(data.type & PERFCOUNTER_LONGRUNAVG));
   if (!(data.type & PERFCOUNTER_U64))
     return;
-  data.u64 -= amt;
+  if (amt > data.u64) { // would wrap
+    assert(false); // Assert in debug builds
```
Comment (tchaikov) [inline src/common/perf_counters.cc:194]:
```c++
assert(data.u64 >= amt);
```
for better readability.

File: src/common/perf_counters.cc [hunk 3]
```diff
@@ -190,7 +190,14 @@ void PerfCounters::dec(int idx, uint64_t amt)
...
+    lderr(m_cct) << "WARNING: PerfCounters::dec() " << data.u64 << " - "
```
Comment (tchaikov) [inline src/common/perf_counters.cc:195]: might want to use `__func__` ?

File: src/common/perf_counters.cc [hunk 4]
```diff
@@ -190,7 +190,14 @@ void PerfCounters::dec(int idx, uint64_t amt)
...
+            << amt << " would wrap. Setting to 0" << dendl;
```
Comment (tchaikov) [inline src/common/perf_counters.cc:196]: how about printing out `idx` as well, so we understand which the actual counter is where we might miss a `inc()` or `set()` call?

File: src/common/perf_counters.cc (hunk 2 above)
Comment (badone) [reply src/common/perf_counters.cc]: ack

File: src/common/perf_counters.cc (hunk 1 above)
Comment (badone) [reply src/common/perf_counters.cc]: ack

File: src/common/perf_counters.cc (hunk 3 above)
Comment (badone) [reply src/common/perf_counters.cc]: __func__ outputs just 'dec' which I thought was less informative than 'PerfCounters::dec()' but sure :)

**Review by cbodley — APPROVED** [review]

**Review by tchaikov — APPROVED** [review]
s/perf_conters/perf_counters/

**Comment by badone** [comment]
> s/perf_conters/perf_counters/

Damn! Copied 37ff97523d45788af74b4d8986ed116c22367ce5 :(

**Comment by badone** [comment]
Removing my test modification since that will cause the test to assert every time. The ceph API tests are failing now as well due to the assert. Looks like we let counters go negative in many and varied places so this change is valid IMHO.

File: src/common/perf_counters.cc [hunk 5]
```diff
@@ -190,7 +190,15 @@ void PerfCounters::dec(int idx, uint64_t amt)
   ceph_assert(!(data.type & PERFCOUNTER_LONGRUNAVG));
   if (!(data.type & PERFCOUNTER_U64))
     return;
-  data.u64 -= amt;
+  if (amt > data.u64) { // would wrap
+    assert(data.u64 >= amt); // Assert in debug builds
```
Comment (tchaikov) [inline src/common/perf_counters.cc:194]: seems we cannot enable this `assert()` yet
```
2020-10-30T04:53:22.460 INFO:teuthology.orchestra.run.smithi117:> sudo ceph config set mgr mgr/devicehealth/enable_monitoring false --force
2020-10-30T04:53:22.677 INFO:teuthology.orchestra.run.smithi117.stderr:python3.6: /build/ceph-16.0.0-6749-g874c7de6/src/common/perf_counters.cc:194: void ceph::common::PerfCounters::dec(int, uint64_t): Assertion `data.u64 >= amt' failed.
2020-10-30T04:53:22.679 DEBUG:teuthology.orchestra.run:got remote process result: None
2020-10-30T04:53:22.679 ERROR:teuthology.contextutil:Saw exception from nested tasks
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/contextutil.py", line 31, in nested
    vars.append(enter())
  File "/usr/lib/python3.6/contextlib.py", line 81, in __enter__
    return next(self.gen)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2020-10-29-2343/qa/tasks/ceph.py", line 1412, in run_daemon
    remote.run(args=cmd.split(' '))
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/remote.py", line 215, in run
    r = self._runner(client=self.ssh, name=self.shortname, **kwargs)
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/run.py", line 446, in run
    r.wait()
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/run.py", line 160, in wait
    self._raise_for_status()
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/orchestra/run.py", line 178, in _raise_for_status
    raise CommandCrashedError(command=self.command)
teuthology.exceptions.CommandCrashedError: Command crashed: 'sudo ceph config set mgr mgr/devicehealth/enable_monitoring false --force'
```

see https://pulpito.ceph.com/kchai-2020-10-30_02:27:32-rados-wip-kefu-testing-2020-10-29-2343-distro-basic-smithi/5572986/


and https://pulpito.ceph.com/kchai-2020-10-30_02:27:32-rados-wip-kefu-testing-2020-10-29-2343-distro-basic-smithi/

**Review by tchaikov — CHANGES_REQUESTED** [review]

File: src/common/perf_counters.cc (hunk 5 above)
Comment (badone) [reply src/common/perf_counters.cc]: OK, but I'm just going to comment it out as we should be able to have this assert at some time in the future. It is an unsigned type after all.

**Review by tchaikov — CHANGES_REQUESTED** [review]
```
2020-11-06T17:23:18.389 INFO:tasks.cram.client.0.smithi082.stdout:/home/ubuntu/cephtest/archive/cram.client.0/misplaced.t: failed
2020-11-06T17:23:18.400 INFO:tasks.cram.client.0.smithi082.stdout:--- misplaced.t
2020-11-06T17:23:18.401 INFO:tasks.cram.client.0.smithi082.stdout:+++ misplaced.t.err
2020-11-06T17:23:18.401 INFO:tasks.cram.client.0.smithi082.stdout:@@ -1,30 +1,129 @@
2020-11-06T17:23:18.401 INFO:tasks.cram.client.0.smithi082.stdout:   $ ceph balancer off
2020-11-06T17:23:18.401 INFO:tasks.cram.client.0.smithi082.stdout:+  2020-11-06T17:22:19.857+0000 7fac25670700 -1 WARNING: dec 0 - 1 for counter 94006 would wrap. Setting to 0
2020-11-06T17:23:18.402 INFO:tasks.cram.client.0.smithi082.stdout:+  2020-11-06T17:22:19.857+0000 7fac25670700 -1 WARNING: dec 0 - 1 for counter 94006 would wrap. Setting to 0
2020-11-06T17:23:18.402 INFO:tasks.cram.client.0.smithi082.stdout:+  2020-11-06T17:22:19.857+0000 7fac25670700 -1 WARNING: dec 0 - 1 for counter 94006 would wrap. Setting to 0
2020-11-06T17:23:18.402 INFO:tasks.cram.client.0.smithi082.stdout:+  2020-11-06T17:22:20.030+0000 7fac25670700 -1 WARNING: dec 0 - 1 for counter 94006 would wrap. Setting to 0
2020-11-06T17:23:18.402 INFO:tasks.cram.client.0.smithi082.stdout:+  2020-11-06T17:22:20.030+0000 7fac25670700 -1 WARNING: dec 0 - 1 for counter 94006 would wrap. Setting to 0
2020-11-06T17:23:18.403 INFO:tasks.cram.client.0.smithi082.stdout:+  2020-11-06T17:22:20.030+0000 7fac25670700 -1 WARNING: dec 0 - 1 for counter 94006 would wrap. Setting to 0
2020-11-06T17:23:18.403 INFO:tasks.cram.client.0.smithi082.stdout:   $ ceph balancer mode none
2020-11-06T17:23:18.403 INFO:tasks.cram.client.0.smithi082.stdout:+  2020-11-06T17:22:20.129+0000 7ff3ce09f700 -1 WARNING: dec 0 - 1 for counter 94006 would wrap. Setting to 0
2020-11-06T17:23:18.403 INFO:tasks.cram.client.0.smithi082.stdout:+  2020-11-06T17:22:20.129+0000 7ff3ce09f700 -1 WARNING: dec 0 - 1 for counter 94006 would wrap. Setting to 0
2020-11-06T17:23:18.403 INFO:tasks.cram.client.0.smithi082.stdout:+  2020-11-06T17:22:20.129+0000 7ff3ce09f700 -1 WARNING: dec 0 - 1 for counter 94006 would wrap. Setting to 0
2020-11-06T17:23:18.403 INFO:tasks.cram.client.0.smithi082.stdout:+  2020-11-06T17:22:20.304+0000 7ff3ce09f700 -1 WARNING: dec 0 - 1 for counter 94006 would wrap. Setting to 0
2020-11-06T17:23:18.403 INFO:tasks.cram.client.0.smithi082.stdout:+  2020-11-06T17:22:20.304+0000 7ff3ce09f700 -1 WARNING: dec 0 - 1 for counter 94006 would wrap. Setting to 0
2020-11-06T17:23:18.404 INFO:tasks.cram.client.0.smithi082.stdout:+  2020-11-06T17:22:20.304+0000 7ff3ce09f700 -1 WARNING: dec 0 - 1 for counter 94006 would wrap. Setting to 0
```

see https://pulpito.ceph.com/kchai-2020-11-06_15:51:09-rados-wip-kefu-testing-2020-11-06-1936-distro-basic-smithi/5595811/

**Comment by ljflores** [comment]
jenkins retest this please

**Comment by badone** [comment]
Hey @ljflores I'm going to set DNM on this for now (should have been set already) and I'll discuss this with you when I get back from PTO.

**Comment by badone** [comment]
This was fixed in https://tracker.ceph.com/issues/66231

**Pull request closed without merging**